### PR TITLE
SWIK-1856 Record presentation rooms

### PR DESCRIPTION
**Pull request by rmeissn**
This branch adds to issue data (created by corresponding branch of slidewiki-platform) to the service and to produce a video out of it. Needs to be tested together with the mentioned branch of slidewiki-platform.

Questions: 
* Am I removing all temporarily created files?
* Are any try/catch or promise/catch statements missing?
* Will the email workflow work?
* Should we add a cron job to remove videos after some time or delete them after first download or keep them without a time limit?
* Is it useful to keep picture/duration list and audio track for other purposes?

NOTE: Github diff view shows way more differences than expected - e.g. storeThumbnail shouldn't have so much differences....

### CODE REVIEW

File: application/routes.js [hunk 1]
```diff
@@ -378,6 +378,101 @@ module.exports = function(server) {
...
+        maxBytes: 10485760, //100MB
+        parse: true,
+        allow: 'multipart/form-data',
+        failAction: 'log'
+      },
+      validate: {
+        options: { convert: true },
+        payload: Joi.object({
+          audioFile: Joi.object().required(),
+          slideTimings: Joi.string().trim().min(5).required()
```
Comment (TBoonX) [inline application/routes.js]: Please add as comment or in the route description what are the requirements (like why slideTimings needs to be at least 5 pparts)

File: application/controllers/handler.js [hunk 2]
```diff
@@ -197,31 +174,76 @@ let handlers = module.exports = {
...
+    let mimeType = request.payload.audioFile.filename.toLowerCase().includes('webm') ? 'audio/webm' : 'audio/ogg';
+    let currentDate = new Date().getTime();
+    let path = '/tmp/' + currentDate + '/';
+    let pictureListName = 'pics.txt';
+    let audioTrackName = 'audioTrack' + ((mimeType === 'audio/webm') ? '.webm' : '.ogg');
+    let outputName = 'D' + request.query.deckID + 'R' + request.query.revision + '-' + currentDate + '.mp4';
+    fs.mkdirSync(path);
+    fs.copyFileSync(request.payload.audioFile.path, path + audioTrackName);
+    fs.unlinkSync(request.payload.audioFile.path);
+    let slideTimes = JSON.parse(request.payload.slideTimings);// let slideTimes = { 1518009573009: '45070-2', 1518009576983: '45071-2', 1518009581167: '45072-3', 1518009587152: '45072-3' };
```
Comment (TBoonX) [inline application/controllers/handler.js]: try catch is missing

File: application/controllers/handler.js [hunk 3]
```diff
@@ -197,31 +174,76 @@ let handlers = module.exports = {
...
+
+    let pics = await downloadSlidePictures(slideList, 1920, path);
+    createFFmpegTimingsFile(pics, slideTimes, path, pictureListName);
+    try {
+      let exec =  require('util').promisify(child.exec);
+      await exec('ffmpeg -f concat -safe 0 -i ' + path + pictureListName + ' -i ' + path + audioTrackName + ' -vsync cfr -c:v libx264 -tune stillimage -c:a aac -b:a 64k ' + path + outputName);
+      fs.copyFileSync(path + outputName, require('../configuration').fsPath + '/videos/' + outputName);
+      // fs.unlinkSync(path + outputName);
+      console.log('Finished Video, sending mail');
+      sendVideoMail(request.auth.credentials.userid, request.query.deckID, request.query.revision, currentDate, outputName);
```
Comment (TBoonX) [inline application/controllers/handler.js]: TypeError: Cannot read property 'userid' of null

**Comment by TBoonX** [comment]
_Will the email workflow work?_
Should. Perhaps we have to change the email text ...

_Should we add a cron job to remove videos after some time or delete them after first download or keep them without a time limit?_
I propose to remove videos after 2 weeks. This info has to be visible in the platform UI and  the email from the user service.

_Is it useful to keep picture/duration list and audio track for other purposes?_
Not from where I standing.

**Comment by rmeissn** [comment]
@TBoonX haven't yet digged into the cron job, but fixed all the other requests

File: application/controllers/handler.js [hunk 4]
```diff
@@ -231,6 +263,64 @@ let handlers = module.exports = {
...
+  fs.writeFileSync(path + pictureListName, toPrint.join('\n'));
+}
+
+function sendVideoMail(userID, deckID, revision, date, fileName) {
+  return rp.post({
+    uri: Microservices.user.uri + '/user/' + userID + '/sendEmail',//userid
+    body: {
+      reason: 2,//new video
+      data: {fileName: fileName, creationDate: date, deck: deckID, revision: revision}
+    },
```
Comment (TBoonX) [inline application/controllers/handler.js:316]: You have to set the jwt in the header

Comment (TBoonX) [reply application/controllers/handler.js:316]: I fixed it. user service throws an error because there is no mail server.

**Review by TBoonX — APPROVED** [review]
Works

**Comment by TBoonX** [comment]
As long as videos are not removed automatically, please assign someone who is responsible for deleting them manually from time to time.

**Comment by rmeissn** [comment]
I will add CRON stuff to this branch as well. I wouldn't count on that someone is looking into this manually.

File: application/controllers/handler.js (hunk 4 above)
Comment (rmeissn) [reply application/controllers/handler.js:316]: Many thanks! I guess it'll just work

**Comment by rmeissn** [comment]
@TBoonX I've added a cron job and tested it locally. Please have a look at the latest commit. If it's okay I consider this PR as finished and ready to merge.

**Comment by TBoonX** [comment]
Cron stuff looks good, cleanup script works. Ready to be merged!

**Pull request merged**